daemon/mgr/image_store.go

```diff
@@ -0,0 +1,280 @@
+func (store *imageStore) GetPrimaryReference(ref reference.Named) (reference.Named, error) {
+	trimRef := reference.TrimTagForDigest(ref)
+
+	store.Lock()
+	defer store.Unlock()
+
+	if p, ok := store.primaryRefIndexByRef[trimRef.String()]; ok {
+		return p, nil
+	}
+	return nil, errtypes.ErrNotfound
+}
+
+// Search returns the imageID, reference.Named by the given reference.Named.
+//
+// NOTE: the reference.Named in the result maybe different from the given reference.Named
+// because it can be added latest tag if missing.
+func (store *imageStore) Search(ref reference.Named) (digest.Digest, reference.Named, error) {
+	store.Lock()
+	defer store.Unlock()
+
+	trimRef := reference.TrimTagForDigest(ref)
+
+	// if the reference is searchable tagged or digest reference
+	if p, ok := store.primaryRefIndexByRef[trimRef.String()]; ok {
+		return store.idIndexByPrimaryRef[p.String()], ref, nil
+	}
+
+	// try to add latest if the reference is only name without tag or digest
+	if reference.IsNamedOnly(ref) {
+		latestRef := reference.WithDefaultTagIfMissing(ref)
+		if p, ok := store.primaryRefIndexByRef[latestRef.String()]; ok {
+			return store.idIndexByPrimaryRef[p.String()], latestRef, nil
+		}
+	}
+
+	// if the reference is short ID or ID
+	//
+	// NOTE: by default, use the sha256 as the digest algorithm
+	id, err := store.searchIDs(digest.Canonical.String(), ref.String())
+	if err != nil {
+		return "", nil, err
+	}
+	return id, ref, nil
+}
+
+func (store *imageStore) searchIDs(algo string, prefixID string) (digest.Digest, error) {
+	var ids []digest.Digest
+
+	fn := func(_ patricia.Prefix, item patricia.Item) error {
+		if got, ok := item.(digest.Digest); ok {
+			ids = append(ids, got)
+		}
+
+		if len(ids) > 1 {
+			return pkgerrors.Wrap(errtypes.ErrTooMany, "image: "+prefixID)
+		}
+		return nil
+	}
+
+	if err := store.idSet.VisitSubtree(patricia.Prefix(algo+":"+prefixID), fn); err != nil {
+		return "", err
+	}
+
+	if len(ids) == 0 {
+		return "", pkgerrors.Wrap(errtypes.ErrNotfound, "image: "+prefixID)
+	}
+	return ids[0], nil
+}
+
+// AddReference adds new reference to the imageID.
+func (store *imageStore) AddReference(id digest.Digest, primaryRef reference.Named, ref reference.Named) error {
+	if reference.IsNamedOnly(ref) ||
+		reference.IsNamedOnly(primaryRef) {
+
+		return pkgerrors.Wrap(errtypes.ErrInvalidType, "invalid reference: missing a tag or digest")
+	}
+
+	var (
+		trimRef        = reference.TrimTagForDigest(ref)
+		trimPrimaryRef = reference.TrimTagForDigest(primaryRef)
+	)
+
+	store.Lock()
+	defer store.Unlock()
+
+	// remove the relationship if the ref has been used by other
+	if oldP, ok := store.primaryRefIndexByRef[trimRef.String()]; ok {
+		// if the trimRef is primary reference
+		if oldP.String() == trimRef.String() {
+			if oldP.String() != trimPrimaryRef.String() {
+
+				return pkgerrors.Wrap(errtypes.ErrInvalidType, "invalid reference: cannot replace primary reference")
```

Comment:
The following command
```sh
pouch pull registry.hub.docker.com/library/busybox@sha256:552644f12d0cb92449213f701c88fabf8e5a7acfae3fcefd144b592293ded713

pouch pull busybox:1-uclibc
```
will make the pouchd report error like:
```sh
ERRO[2018-05-07T16:22:07.055029807+08:00] failed to pull image busybox:1-uclibc:1-uclibc: invalid reference: cannot replace primary reference: invalid type
```

From my point of view, if a Ref has already been a PrimaryRef, we may not need to maintain its relationship with other PrimaryRefs.

So maybe it is more reasonable that just `return nil` other then error here.

@fuweid WDYT?

Reply:
@YaoZengzeng , I think we should maintain the relationship.
If not, the searchable reference can override the primary relationship. In this case, we cannot clean the image in the containerd side. It maybe cause the dirty data in the host. WDYT?

I have updated my code. In this case, we should check the reference before we add it. Please take a look.

Thanks for the suggestion.